src/full/Agda/Compiler/Malfunction.hs

```diff
@@ -0,0 +1,247 @@
+          $$ sect "Treeless (abstract syntax)"    (text . show $ tt)
+          $$ sect "Treeless (concrete syntax)"    (pretty tt)
+    _ -> return ()
+    where
+      sect t dc = text t $+$ nest 2 dc $+$ text ""
+      header c h = let cs = replicate 15 c
+                   in text $ printf "%s %s %s" cs h cs
+      q = defName def
+      defntype = case theDef def of
+        Constructor{}  -> "constructor"
+        Primitive{}    -> "primitive"
+        Function{}     -> "function"
+        Datatype{}     -> "datatype"
+        Record{}       -> "record"
+        AbstractDefn{} -> "abstract"
+        Axiom{}        -> "axiom"
+
+-- TODO: Maybe we'd like to refactor this so that we first do something like
+-- this (in the calling function)
+--
+--    partition [Definition] -> ([Function], [Primitive], ...)
+--
+-- And then we can simply do the topologic sorting stuff on the functions only.
+-- This would certainly make this funciton a helluva lot cleaner.
+--
+-- | Compiles a whole module
+mlfMod
+  :: [Definition]   -- ^ All visible definitions
+  -> TCM Mod
+mlfMod allDefs = do
+  grps' <- mapMaybeM act allDefs
+  let (primsAndAxioms, tlFunBindings) = partitionEithers grps'
+      (prims, axioms) = partitionEithers primsAndAxioms
+  env <- getCompilerEnv (getConstructors allDefs) tlFunBindings
+  let (MMod funBindings ts) = compile env tlFunBindings
+      primBindings = catMaybes $ Mlf.runTranslate (mapM (uncurry Mlf.compilePrim) prims) env
+      axiomBindings = catMaybes $ Mlf.runTranslate (mapM Mlf.compileAxiom axioms) env
+  return $ MMod (axiomBindings ++ primBindings ++ funBindings) ts
+    where
+      act :: Definition -> TCM (Maybe (Either (Either (QName, String) QName) (QName, TTerm)))
+      act def@Defn{defName = q, theDef = d} = case d of
+        Function{}                -> fmap Right <$> getBindings def
+        Primitive{ primName = s } -> return $ Just $ Left $ Left (q, s)
+        Axiom{}                   -> return $ Just $ Left $ Right q
+        _                         -> return Nothing
+
+compile :: Mlf.Env -> [(QName, TTerm)] -> Mod
+compile env bs = Mlf.compile env bs
+
+qnamesInTerm :: Set QName -> TTerm -> Set QName
+qnamesInTerm s t = go t s
+  where
+    go :: TTerm -> Set QName -> Set QName
+    go t qs = case t of
+      TDef q           -> Set.insert q qs
+      TApp f args      -> foldr go qs (f:args)
+      TLam b           -> go b qs
+      TCon q           -> Set.insert q qs
+      TLet a b         -> foldr go qs [a, b]
+      TCase _ _ p alts -> foldr qnamesInAlt (go p qs) alts
+      _                -> qs
+      where
+        qnamesInAlt a qs = case a of
+          TACon q _ t -> Set.insert q (go t qs)
+          TAGuard t b -> foldr go qs [t, b]
+          TALit _ b   -> go b qs
+
+
+-- | The argument allNames is optional. If you provide an empty list concrete
+-- names will not be appended to the NameId
+getCompilerEnv :: [[QName]] -> [(QName, TTerm)] -> TCM Mlf.Env
+getCompilerEnv allcons bs
```

Comment:
This adds all constructors used in ``bs`` to the returned ``Env``, right? Maybe it would be easier to lift your ``translateDefM`` into the ``TCM`` Monad, then you could directly use ``TCM`` and would not need to precompute your environment?